### ISSUE TITLE
Using the aggregate() function with a filter as a default value returns a NULL when new features are created

### ISSUE THREAD

**Issue by raherin** [qgis/QGIS#59740]
### What is the bug or the crash?

In QGIS versions 3.34 and later, the behaviour of the aggregate function has changed so that a NULL value is returned when a new feature is created where the function is used as a default value with a filter set. In version 3.32, the function as used would return a value according to the formula settings.

### Steps to reproduce the issue

To reproduce the issue, create a simple point layer with two fields using the following settings:

| *Field* | *Type* | *Widget settings* | *Default value* |
| --------- | -------- | ----------------- | ------------------------------|
| group | STRING | _Value map_ with values 'A, B, and C' | 'A' (optional) |
| pt_num| INT | _Range_ with minimum value '1' and step '1' | Use expression `aggregate('survey_pts','max',"pt_num",filter:= "group" = attribute(@parent, 'group')) + 1` with 'Apply default value on update' enabled|

The desired behaviour of the function is that each new feature created would result in a default value for _pt_num_ that increases in value based on the highest value for that field compared to other features with the same value in the _group_ field. The first feature for each _group_ would have to have _pt_num_ set to '1' or you could wrap the _aggregate()_ function in a _coalesce()_ function (`coalesce(aggregate('survey_pts','max',"pt_num",filter:= "group" = attribute(@parent, 'group')) + 1, 1)`, but I want to focus on the behaviour of the aggregate() function.

When using this setup in QGIS 3.32, the behaviour is as expected. Each new feature created gets an increasing value for _pt_num_ when it is created based on the value selected for the _group_ field:
<img width="1082" alt="image" src="https://github.com/user-attachments/assets/9b8bcdd5-fba5-4dd2-8878-cf92551f18ae">

 However, if the project is created in QGIS 3.34 or higher, the default value for _pt_num_ is returned as _NULL_:
<img width="1192" alt="image" src="https://github.com/user-attachments/assets/75bfc295-bc06-46c6-9c12-621a94a8e8db">

In either of the QGIS versions, the formula will apply when you open the attribute table and edit any of the existing features, the _NULL_ value is only returned when a new feature is created.

### Versions

I have tested this setup in QGIS versions 3.32, 3.34 and 3.38 and the formula only works as expected in version 3.32

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

**Comment by raherin** [comment]
I should also clarify that it makes a difference which version of QGIS the project is created in. If I create the project in v3.32 and then run it in v3.34, it will work as expected. Also if I create the project in v3.34 and run it in v3.32 it will work as expected. So it is like this:

| Version created | Works in v3.32 | Works in v3.34 |
| ----- | ----- | ----- |
| 3.32 | yes | yes |
| 3.34 | yes | no |

I've also attached example projects created in 3.34 and 3.32


[aggregate_func-3.32.zip](https://github.com/user-attachments/files/18009200/aggregate_func-3.32.zip)

[aggregate_func-3.34.zip](https://github.com/user-attachments/files/18009204/aggregate_func-3.34.zip)

**Comment by uclaros** [comment]
After changing the value in the _group_ combobox, it seems that the correct result is returned, at least in master.

**Comment by raherin** [comment]
@uclaros yes, that is the behaviour I get as well. However, there may be use cases where the field being used as the filter is not editable, for example it may be a field referencing a table relation, so it would be preferred to return the correct value by default.
